### PR TITLE
Build: divide each test run into 5 deterministic separate jobs

### PR DESCRIPTION
**Pull request by frenchy64**
## Purpose and Motivation

Related https://github.com/supercollider/supercollider/issues/6439

This PR speeds up CI runs of `test-macos` and `test-linux` by about 3-4x (12m => 4m).

The approach is to deterministically partition the tests into separate concurrent jobs. Test failures will be locally reproducible by passing the same arguments to the test script as whichever job fails.

Demonstration: https://github.com/frenchy64/supercollider/actions/runs/11807687690

## Types of changes

- CI Build performance improvement

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review 

## Merge notes

Reviewers please help me verify that the test partitioning logic is deterministic (per commit). Each test should run exactly once and in the same order.

`.clumps()` partitions the tests but it assumes the array it is clumping is itself deterministic. Please double check this.

Suggestions welcome on how to improve `testsuite/scripts/setup_test_clumps_output.sh`. Generating JSON from bash is ugly, but it at least seems lean and fast.

### CODE REVIEW

File: testsuite/scripts/sclang_test_runner.scd [hunk 1]
```diff
@@ -144,6 +167,10 @@ if(File.exists(~testProto).not) {
 	// ~writeResult.(tests, file);
 	~verboseSetup.if({"After exclude: ".post; tests.do(_.postln)});
 
+  tests = ~selectThisTestClump.(tests, ~thisClump, ~nClumps);
```
Comment (capital-G) [inline testsuite/scripts/sclang_test_runner.scd:170]: Is the indentation off here? Looks strange on GH.

**Review by capital-G — COMMENTED** [review]
Hey, thanks for this PR, this is really great!

I opposed parallelizing the tests in the pasts, but running them in parallel on dedicated machines looks like a good solution to me.
I found everything well implemented and clearly documented and it really benefits the pipeline - so from my side this is good to merge except the one small comment.

But let's wait what @dyfer thinks about it.

**Comment by dyfer** [comment]
Hi @frenchy64 and thanks for the PR!
I generally think splitting the tests to speed them up is great.

> `.clumps()` partitions the tests but it assumes the array it is clumping is itself deterministic. Please double check this.

Good point. I'm not 100% sure about this. I think technically there's no guarantee that the class tree is deterministic, as it uses an `IdentitySet`... maybe we should sort the test classes before clumping them?

Regarding the Bash script: I appreciate the automation, but maybe for transparency it would be easier to just use the build matrix for this, something like
```
matrix:
        this-clump: [0, 1, 2, 3, 4] # consecutive numbers starting from 0
        n-clumps: 5 # needs to be one larger than the highest `this-clump`
```
This of course is error-prone, but it's also very transparent and avoids passinge JSON and hidden automation.

I think we need to 
1. test/ensure the deterministic list
2. consider using Bash vs manual matrix

File: testsuite/scripts/sclang_test_runner.scd (hunk 1 above)
Comment (dyfer) [reply testsuite/scripts/sclang_test_runner.scd:170]: Yes, please fix the indentation @frenchy64

**Review by dyfer — CHANGES_REQUESTED** [review]
see previous comment

**Comment by frenchy64** [comment]
> EDIT: maybe the number of clumps could be calculated with ${{ strategy.job-total }} instead of specifying it manually, but I haven't tested that.

Seems to work great! Halved the lines of changes needed.

**Comment by dyfer** [comment]
I like it!

Before merging this I'd like to consider some aesthetic aspects:

- should the test-clump parameter had a short comment saying that it divides the test suite into a number of segments equal to the number of items in the array, and also that it doesn't matter when what these numbers are?
- I wonder if in the test running and/or test result steps we should add posting like "test suite segment X out of X"?



I wonder if it's worth it to add some

**Comment by frenchy64** [comment]
Yes, good ideas. I will add that. I also need to investigate the sorting/determinism before this is merged.

**Comment by capital-G** [comment]
One thing that should be verified before merging - do indeed all tests run as expected in CI?

I recently thought about writing a JUnit XML wrapper/generator for our test suite, it seems like such a thing would come handy now when we split the run across multiple machines.

File: testsuite/scripts/sclang_test_runner.scd [hunk 2]
```diff
@@ -1,15 +1,32 @@
...
+		~nClumps = 1;
+	};
 	~exitWhenDone = true;
 } {
 	~testProto = PathName(thisProcess.nowExecutingPath).pathOnly +/+ "test_run_proto_gha.scd";
 	// ~testProto = PathName(thisProcess.nowExecutingPath).pathOnly +/+ "test_run_proto_all.scd"; // for testing
 	~testResult = PathName(thisProcess.nowExecutingPath).pathOnly +/+ "run" +/+ "test_result.scxtar";
+	~thisClump = 0;
+	~nClumps = 1;
 };
```
Comment (dyfer) [inline testsuite/scripts/sclang_test_runner.scd:30]: Quick suggestion: instead of two definitions of 
```supercollider
~thisClump = 0;
~nClumps = 1;
```
you could add below this block:
```supercollider
~thisClump ?? {~thisClump = 0};
~nClumps ?? {~nClumps = 1};
```
or
```supercollider
~thisClump = ~thisClump ? 0;
~nClumps = ~nClumps ? 1;
```

**Comment by dyfer** [comment]
> One thing that should be verified before merging - do indeed all tests run as expected in CI?

This is a good point.

> I recently thought about writing a JUnit XML wrapper/generator for our test suite, it seems like such a thing would come handy now when we split the run across multiple machines.

However, I'm weary of using a wrapper for a testsuite, as it provides hurdles when running tests locally (point in case: qpm). 

I think we could consider adding the script which generates the list of all tests after the tests run and compares the combined results between the jobs (exchanging that information would be cumbersome, but definitely doable).

**Comment by capital-G** [comment]
> However, I'm weary of using a wrapper for a testsuite, as it provides hurdles when running tests locally (point in case: qpm).

My idea is to output the results of the test suit (also) as JUnit XML format so it can be parsed by tools such as https://github.com/marketplace/actions/publish-test-results  - it wouldn't affect the running of the tests, it is simply a standard to serialize unit test results. But lets keep this in its separate Issue/PR (but if you have already objections to this dyfer plz write me)

**Comment by frenchy64** [comment]
> I think we could consider adding a script which generates the list of all tests after the tests run and compares it with the results combined from all the jobs (exchanging that information would be cumbersome, but definitely doable).

I've done this successfully several times, I can outline how to achieve this using GitHub Actions.

Here's an example of a job that downloads prior test results, then fans-out tests and collects results after fanning-in, then prints and uploads the test results to load-balance future runs.

1. [Download previous timings via a cache and upload an artifact to share with splits](https://github.com/threatgrid/ctia/blob/3478e16c2f19ccdbd9d2cd2e3957ec67cd15b867/.github/workflows/pr.yml#L82-L114)
2. [Download the timings artifact on each split](https://github.com/threatgrid/ctia/blob/3478e16c2f19ccdbd9d2cd2e3957ec67cd15b867/.github/workflows/pr.yml#L240-L244)
3. [Upload results from each split](https://github.com/threatgrid/ctia/blob/3478e16c2f19ccdbd9d2cd2e3957ec67cd15b867/.github/workflows/pr.yml#L302-L307)
4. [Compile all results and upload to cache for future builds](https://github.com/threatgrid/ctia/blob/3478e16c2f19ccdbd9d2cd2e3957ec67cd15b867/.github/workflows/pr.yml#L340-L365)

The logic on how to determine which tests to run is [here](https://github.com/threatgrid/ctia/blob/3478e16c2f19ccdbd9d2cd2e3957ec67cd15b867/dev/ctia/dev/split_tests.clj#L102-L157).

**Comment by smoge** [comment]
Thanks for the excellent contribution.

Your PR introduces deterministic partitioning to predictably distribute tests across parallel CI jobs, which helps with the reproducibility of test runs locally. 

However, I noticed there might be old confusion in this community regarding terminology—this kind of determinism (predictable test distribution) is different from property-based testing determinism (where test case generation follows well-defined procedures that produce consistent results). Property-based tests are inherently deterministic by design.

 Making this distinction clear could help avoid confusion. These are two separate aspects of test reproducibility, and there is still a lot of confusion in the community regarding the word "determinism" in the context of tests.
 
 Thanks again

**Comment by frenchy64** [comment]
I'm going to continue work [here](https://github.com/frenchy64/supercollider/pull/7) where I have permission to cancel broken builds and I'll push my work here once it works.

**Comment by smoge** [comment]
> I'm going to continue work [here](https://github.com/frenchy64/supercollider/pull/7)

Maybe this can be developed elsewhere, but to mention something pertinent to supercollider testing enhancement perspectives.

SuperCollider's scheduler has to handle concurrent operations (even though it's single-core, it still needs to handle multiple "concurrent" tasks). That is, sclang is a single-core, non-deterministic scheduler. It needs to handle timing and order of operations, race conditions, and try the best possible correct sequencing.

There is a compelling development of using a deterministic parallel scheduler to allow you to consistently reproduce timing-related bugs using the same seed. Also, that's the only safe way to shrink: when finding bugs, you could reliably shrink test cases to minimal examples since the scheduling would remain consistent. Also, one way to fix tricky bugs regarding edge-case with race conditions (an increasingly common bug when using large GUI systems or doing something non-trivial)

https://stevana.github.io/parallel_property-based_testing_with_a_deterministic_thread_scheduler.html

**Comment by capital-G** [comment]
Thanks for your input regarding property based testing @smoge, but I think this PR has nothing to do with property-based testing as it only changes the way we separate and distribute tests on our test runners in CI.
The determinism discussed here is about the partition of said tests across the runners and not about any determinism within tests of the language or server.

If you want to discuss the introduction of property testing please do so in a separate issue/PR/RFC to keep things focused, atomic and isolated.

**Comment by smoge** [comment]
> I think this PR has nothing to do with property-based testing as this PR only changes the way we separate and distribute tests on our test runners in CI.

I know that.No problems, @capital-G. If this issue needs some more attention elsewhere, I'm available to contribute. Take care.

**Comment by frenchy64** [comment]
I found a bug in my partitioning algorithm where some tests were being dropped (as speculated by @capital-G). The problem was I was making the clump size too small when certain non-divisible numbers of tests were provided, so an extra clump was being made for the leftovers that no job was picking up.

I'm hardening the code and adding assertions to double and triple check the partitions are correct.

**Comment by smoge** [comment]
> I found a bug in my partitioning algorithm where some tests were being dropped (as speculated by @capital-G). The problem was I was making the clump size too small when certain non-divisible numbers of tests were provided, so an extra clump was being made for the leftovers that no job was picking up.
> 
> I'm hardening the code and adding assertions to double and triple check the partitions are correct.

I quickly looked at ~selectThisTestClump and drafted some code using ceiling division just to check it.

```
~selectThisTestClump = {|tests,thisClump,nClumps|
    tests.sort.clump(tests.size()/nClumps).at(thisClump) ?? #[];
};
```

```supercollder
~selectThisTestClump = {|tests, thisClump, nClumps|
    var sorted, clumpSize, clumps, result;
    
    sorted = tests.sort { |a, b| 
        if(a[\suite] == b[\suite]) {
            a[\test] < b[\test]
        } {
            a[\suite] < b[\suite]
        }
    };
    
    // ceiling division / no dropping of tests
    clumpSize = (sorted.size / nClumps).ceil.asInteger;
    
    clumps = sorted.clumps(clumpSize);
    
    // double-check partitioning
    protect {
        var totalTests = sorted.size;
        var testInClumps = clumps.flatten.size;
        
        "Test partitioning:".postln;
        "  Total tests: %".format(totalTests).postln;
        "  Number of clumps: %".format(clumps.size).postln;
        "  This clump index: %".format(thisClump).postln;
        "  Tests in this clump: %".format(clumps[thisClump].size).postln;
        
        if(testInClumps != totalTests) {
            "Test partitioning error: % tests in clumps, expected %".format(testInClumps, totalTests).error;
        };
        
        if(clumps.size > nClumps) {
            "Created % clumps, expected <= %".format(clumps.size, nClumps).error;
        };
    };
    
    // Returns clump, or empty array 
    clumps[thisClump] ?? #[]
};
```
